client/src/test/scala/io/mediachain/client/InMemoryJournalFollower.scala

```diff
@@ -0,0 +1,108 @@
+package io.mediachain.client
+
+import io.mediachain.protocol.Datastore._
+import io.mediachain.protocol.Transactor.JournalListener
+
+import scala.collection.immutable.SortedMap
+
+class InMemoryJournalFollower(val datastore: Datastore) extends JournalListener {
+
+  override def onJournalCommit(entry: JournalEntry): Unit = {
+    // TODO
+  }
+
+  override def onJournalBlock(ref: Reference): Unit =
+    datastore.getAs[JournalBlock](ref).foreach { block =>
+      handleNewBlock(ref, block)
+    }
+
+
+  private var blockRefMap: SortedMap[BigInt, Reference] = SortedMap()
+
+  def blocks: List[JournalBlock] =
+    blockRefMap.values.toList.flatMap(datastore.getAs[JournalBlock])
+
+  def currentBlock: Option[JournalBlock] =
+    blockRefMap
+      .lastOption
+      .flatMap(pair => datastore.getAs[JournalBlock](pair._2))
+
+  // Map of canonical reference to most recent ChainEntry
+  // for that canonical
+  private var chainHeads: Map[Reference, ChainEntry] = Map()
+
+  /// Set of all `CanonicalRecord`s referenced in the blockchain
+  def canonicals: Set[CanonicalRecord] =
+    blocks.toSet.flatMap { block: JournalBlock =>
+      val refs = block.entries.collect {
```

Comment:
you can just write this with the datastore call inside the collect, I think that reads better

```scala
block.entries.collect { block: JournalBlock =>
  val refs = block.entries.collect {
    case e: CanonicalEntry => datastore.getAs[CanonicalRecord](e.ref)
  }
}
```

also, this should be lazy in some way probably? will be easier to reason about if we separate things and see the flow more clearly